contributors/design-proposals/local-storage-pv.md

```diff
@@ -0,0 +1,631 @@
+Pod spec.  There are no special annotations or fields that need to be set in the Pod
+or PVC to distinguish between local and remote storage.  It is abstracted by the
+StorageClass.
+
+#### PersistentVolume Node Affinity
+
+PersistentVolume node affinity is a new concept and is similar to Pod node affinity,
+except instead of specifying which nodes a Pod has to be scheduled to, it specifies which nodes
+a PersistentVolume can be attached and mounted to, influencing scheduling of Pods that
+use local volumes.
+
+For a Pod that uses a PV with node affinity, a new scheduler predicate
+will evaluate that node affinity against the node's labels.  For this initial phase, the
+PV node affinity is only considered by the scheduler for already-bound PVs.  It is not
+considered during the initial PVC/PV binding, which will be addressed in a future release.
+
+Only the `requiredDuringSchedulingIgnoredDuringExecution` field will be supported.
+
+##### API Changes
+
+For the initial alpha phase, node affinity is expressed as an optional
+annotation in the PersistentVolume object.
+
+```
+// AlphaStorageNodeAffinityAnnotation defines node affinity policies for a PersistentVolume.
+// Value is a string of the json representation of type NodeAffinity
+AlphaStorageNodeAffinityAnnotation = "volume.alpha.kubernetes.io/node-affinity"
+```
+
+#### Local volume initial configuration
+
+There are countless ways to configure local storage on a node, with different patterns to
+follow depending on application requirements and use cases.  Some use cases may require
+dedicated disks; others may only need small partitions and are ok with sharing disks.
+Instead of forcing a partitioning scheme on storage administrators, the Local volume
+is represented by a path, and lets the administrators partition their storage however they
+like, with a few minimum requirements:
+
+* The paths to the mount points are always consistent, even across reboots or when storage
+is added or removed.
+* The paths are backed by a filesystem (block devices or raw partitions are not supported for
+the first phase)
+* The storage is mounted according to the provisioner discovery scheme defined below
+* The directories have appropriate permissions for the provisioner to be able to set owners and
+cleanup the volume.
+
+#### Local volume management
+
+Local PVs are statically created and not dynamically provisioned for the first phase.
+To mitigate the amount of time an administrator has to spend managing Local volumes,
+a Local static provisioner application will be provided to handle common scenarios.  For
+uncommon scenarios, a specialized provisioner can be written.
+
+The Local static provisioner will be developed in the
+[kubernetes-incubator/external-storage](https://github.com/kubernetes-incubator)
+repository, and will loosely follow the external provisioner design, with a few differences:
+
+* A provisioner instance needs to run on each node and only manage the local storage on its node.
```

Comment:
The permissions are handled through ClusterRoleBinding.  But I am unaware of a way to have a different ClusterRoleBinding for each instance of a DaemonSet so that it only can see the PVs for that node.

Each provisioner instance will filter the PVs it handles in its code, but the PV informer itself is not filtered.  Each instance gives itself a name of "local-provisioner-<node uid>", and marks every PV It creates with that name in the "provisioned-by" annotation (part of the general [external provisioner](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/volume-provisioning.md)  design).  And it filters the PVs on that annotation.

If we want to filter at the informer level, I could also label the PVs with the provisioner name and filter on labels.  But that's still just relying on the implementation to correctly filter itself.  I would be open to any suggestions you may have.

Reply:
Corrected the above comment to add nodeUID to the name.

Reply:
Is it possible to extend [node authorization](https://kubernetes.io/docs/admin/authorization/node/)?

Reply:
to do that, several things would be required:
* the instance on each node would have to run with the node's credentials (not as a daemonset, etc)
* there would have to be strong ties from a PV to a particular node (not just a label selector or annotation)

at first glance, a division of responsibilities like https://github.com/kubernetes/community/pull/989#discussion_r135700184 seems like it might work better

Reply:
hmm...  do you think this will become a general requirement where daemonset pod can be  restricted to only access local resources, not just for local volume use case?